File: Problems/Algorithms/230. Kth Smallest Element in a BST/KthElementBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<Integer> result;
    
    public int kthSmallest(TreeNode root, int k) {
        result = new ArrayList<>();
        
        aux(root, k);
        
        return result.get(k-1);
    }
    
    private void aux(TreeNode node, int k) {
        if (result.size() >= k || node == null) return;
        
        aux(node.left, k);
        result.add(node.val);
        aux(node.right, k);
    }
}

File: Problems/Algorithms/230. Kth Smallest Element in a BST/KthElementBST.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    private lateinit var result: MutableList<Int>
    
    fun kthSmallest(root: TreeNode?, k: Int): Int {
        result = mutableListOf<Int>()
        
        aux(root, k)
        
        return result[k-1]
    }
    
    private fun aux(node: TreeNode?, k: Int): Unit {
        if (result.size >= k || node == null) return
        
        aux(node?.left, k)
        result.add(node?.`val`)
        aux(node?.right, k)
    }
}

